File: src/main/java/com/kgaft/VulkanLib/Pipelines/PipelineConfiguration/PipelineConfigurer.java
package com.kgaft.VulkanLib.Pipelines.PipelineConfiguration;

import com.kgaft.VulkanLib.Device.LogicalDevice.LogicalDevice;
import com.kgaft.VulkanLib.Pipelines.PipelineConfiguration.PipelineBuilder.*;
import com.kgaft.VulkanLib.Utils.DestroyableObject;
import com.kgaft.VulkanLib.Utils.LwjglObject;
import com.kgaft.VulkanLib.Utils.VkErrorException;
import org.lwjgl.vulkan.*;

import java.lang.instrument.IllegalClassFormatException;
import java.nio.LongBuffer;

import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static org.lwjgl.vulkan.KHRAccelerationStructure.*;
import static org.lwjgl.vulkan.VK13.*;

public class PipelineConfigurer extends DestroyableObject {

    private long pipelineLayout;
    private long descriptorSetLayout;
    private LogicalDevice device;
    private LwjglObject<VkVertexInputBindingDescription.Buffer> inputBindDesc = new LwjglObject<>(VkVertexInputBindingDescription.class, VkVertexInputBindingDescription.Buffer.class, 1);
    private LwjglObject<VkVertexInputAttributeDescription.Buffer> inputAttribDescs;


    public PipelineConfigurer(LogicalDevice device, PipelineBuilder builder) throws IllegalClassFormatException, VkErrorException {
        this.device = device;
        loadDescriptorSetLayout(builder);
        loadPipelineLayout(builder);
        prepareBinding(builder.getVertexInputs());
        prepareInputAttribs(builder.getVertexInputs());
    }

    public long getPipelineLayout() {
        return pipelineLayout;
    }

    public long getDescriptorSetLayout() {
        return descriptorSetLayout;
    }

    public LogicalDevice getDevice() {
        return device;
    }

    public LwjglObject<VkVertexInputBindingDescription.Buffer> getInputBindDesc() {
        return inputBindDesc;
    }

    public LwjglObject<VkVertexInputAttributeDescription.Buffer> getInputAttribDescs() {
        return inputAttribDescs;
    }

    private void loadDescriptorSetLayout(PipelineBuilder endConfiguration) throws IllegalClassFormatException, VkErrorException {
        int capacity = endConfiguration.getUniformBufferInfo().size() +
                endConfiguration.getSamplersInfo().size() +
                endConfiguration.getStorageImagesInfos().size() +
                endConfiguration.getAccelerationStructuresInfos().size() +
                endConfiguration.getStorageBufferInfos().size();
        if (capacity != 0) {
            LwjglObject<VkDescriptorSetLayoutBinding.Buffer> bindings = new LwjglObject<>(VkDescriptorSetLayoutBinding.class, VkDescriptorSetLayoutBinding.Buffer.class,
                    capacity);

            int counter = 0;
            for (UniformBufferInfo uniformBufferInfo : endConfiguration.getUniformBufferInfo()) {
                uboToBind(uniformBufferInfo, bindings.get().get(counter));
                counter++;
            }
            for (SamplerInfo samplerInfo : endConfiguration.getSamplersInfo()) {
                samplerToBind(samplerInfo, bindings.get().get(counter));
                counter++;
            }
            for (AccelerationStructureInfo accelerationStructuresInfo : endConfiguration.getAccelerationStructuresInfos()) {
                asToBind(accelerationStructuresInfo, bindings.get().get(counter));
                counter++;
            }
            for (StorageImageInfo storageImagesInfo : endConfiguration.getStorageImagesInfos()) {
                storageImageToBind(storageImagesInfo, bindings.get().get(counter));
                counter++;
            }
            for (StorageBufferInfo storageBufferInfo : endConfiguration.getStorageBufferInfos()) {
                sboToBind(storageBufferInfo, bindings.get().get(counter));
                counter++;
            }
            LwjglObject<VkDescriptorSetLayoutCreateInfo> layoutInfo = new LwjglObject<>(VkDescriptorSetLayoutCreateInfo.class);
            layoutInfo.get().pBindings(bindings.get());
            long[] res = new long[1];
            VkErrorException.checkVkStatus("Failed to create descriptor set layout: ", vkCreateDescriptorSetLayout(device.getDevice(), layoutInfo.get(), null, res));
            this.descriptorSetLayout = res[0];
        }

    }

    private void loadPipelineLayout(PipelineBuilder endConfiguration) throws IllegalClassFormatException, VkErrorException {
        LwjglObject<VkPushConstantRange.Buffer> pushConstantRanges = new LwjglObject<>(VkPushConstantRange.class, VkPushConstantRange.Buffer.class, endConfiguration.getPushConstantInfos().size());

        int counter = 0;
        for (PushConstantInfo pushConstantInfo : endConfiguration.getPushConstantInfos()) {
            infoToRange(pushConstantInfo, pushConstantRanges.get().get(counter));
            counter++;
        }
        LwjglObject<VkPipelineLayoutCreateInfo> pipelineLayoutInfo = new LwjglObject<>(VkPipelineLayoutCreateInfo.class);
        LongBuffer layouts = LongBuffer.allocate(1);
        if (descriptorSetLayout != 0) {
            layouts.put(descriptorSetLayout);
            layouts.rewind();
            pipelineLayoutInfo.get().pSetLayouts(layouts);
        }

        pipelineLayoutInfo.get().pPushConstantRanges(pushConstantRanges.get());
        long[] layout = new long[1];
        VkErrorException.checkVkStatus("Failed to create pipeline layout: ", vkCreatePipelineLayout(device.getDevice(), pipelineLayoutInfo.get(), null, layout));
    }

    private void prepareBinding(List<VertexInput> inputs) {
        AtomicInteger size = new AtomicInteger();
        inputs.forEach(element -> {
            size.getAndAdd(element.typeSize * element.coordinatesAmount);
        });
        inputBindDesc.get().binding(0);
        inputBindDesc.get().stride(size.get());
        inputBindDesc.get().inputRate(VK_VERTEX_INPUT_RATE_VERTEX);
    }

    private void prepareInputAttribs(List<VertexInput> inputs) throws IllegalClassFormatException {
        int offsetCount = 0;
        int counter = 0;
        inputAttribDescs = new LwjglObject<>(VkVertexInputAttributeDescription.class, VkVertexInputAttributeDescription.Buffer.class, inputs.size());
        for (VertexInput element : inputs) {
            inputAttribDescs.get().get(counter).binding(0);
            inputAttribDescs.get().get(counter).location(element.location);
            inputAttribDescs.get().get(counter).offset(offsetCount);
            inputAttribDescs.get().get(counter).format(element.format);
            offsetCount += element.typeSize * element.coordinatesAmount;
            counter++;
        }
    }


    private void infoToRange(PushConstantInfo info, VkPushConstantRange range) {
        range.size(info.size);
        range.offset(0);
        range.stageFlags(info.shaderStages);
    }

    private void uboToBind(UniformBufferInfo bufferInfo, VkDescriptorSetLayoutBinding result) {
        result.binding(bufferInfo.binding);
        result.descriptorType(VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER);
        result.descriptorCount(bufferInfo.descriptorCount);
        result.stageFlags(bufferInfo.shaderStages);
    }

    private void sboToBind(StorageBufferInfo bufferInfo, VkDescriptorSetLayoutBinding result) {
        result.binding(bufferInfo.binding);
        result.descriptorType(VK_DESCRIPTOR_TYPE_STORAGE_BUFFER);
        result.descriptorCount(bufferInfo.descriptorCount);
        result.stageFlags(bufferInfo.stageFlags);
    }

    private void samplerToBind(SamplerInfo samplerInfo, VkDescriptorSetLayoutBinding result) {
        result.binding(samplerInfo.binding);
        result.descriptorType(VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER);
        result.descriptorCount(samplerInfo.descriptorCount);
        result.stageFlags(samplerInfo.shaderStages);
    }

    private void asToBind(AccelerationStructureInfo asInfo, VkDescriptorSetLayoutBinding result) {
        result.binding(asInfo.binding);
        result.descriptorType(VK_DESCRIPTOR_TYPE_ACCELERATION_STRUCTURE_KHR);
        result.descriptorCount(asInfo.descriptorCount);
        result.stageFlags(asInfo.shaderStages);
    }

    private void storageImageToBind(StorageImageInfo storageImageInfo, VkDescriptorSetLayoutBinding result) {
        result.binding(storageImageInfo.binding);
        result.descriptorType(VK_DESCRIPTOR_TYPE_STORAGE_IMAGE);
        result.descriptorCount(storageImageInfo.descriptorCount);
        result.stageFlags(storageImageInfo.shaderStages);
    }

    @Override
    public void destroy() {
        destroyed = true;
        vkDestroyPipelineLayout(device.getDevice(), pipelineLayout, null);
        vkDestroyDescriptorSetLayout(device.getDevice(), descriptorSetLayout, null);
    }
}

File: src/pipelines/pipeline_config/pipeline_configurer.rs
use ash::vk;
use ash::Device;
use crate::pipelines::pipeline_config::pipeline_builder::*;

pub struct VlPipelineConfigurer {
    device: Device,
    pipeline_layout: vk::PipelineLayout,
    descriptor_set_layout: Option<vk::DescriptorSetLayout>,
    input_bind_desc: vk::VertexInputBindingDescription,
    input_attrib_descs: Vec<vk::VertexInputAttributeDescription>,
}

impl Drop for VlPipelineConfigurer {
    fn drop(&mut self) {
        unsafe { self.destroy() }
    }
}

impl VlPipelineConfigurer {
    pub fn new(device: Device, builder: &VlPipelineBuilder) -> Self {
        let mut configurer = VlPipelineConfigurer {
            device,
            pipeline_layout: vk::PipelineLayout::null(),
            descriptor_set_layout: None,
            input_bind_desc: vk::VertexInputBindingDescription::default(),
            input_attrib_descs: Vec::new(),
        };

        configurer.load_descriptor_set_layout(builder);
        configurer.load_pipeline_layout(builder);
        configurer.prepare_binding(&builder.vertex_inputs);
        configurer.prepare_input_attribs(&builder.vertex_inputs);

        configurer
    }



    fn load_descriptor_set_layout(&mut self, builder: &VlPipelineBuilder) {
        let total_bindings = builder.uniform_buffer_info.len()
            + builder.samplers_info.len()
            + builder.storage_images_info.len()
            + builder.acceleration_structures_info.len()
            + builder.storage_buffer_infos.len();

        let mut bindings = vec![vk::DescriptorSetLayoutBinding::default(); total_bindings];
        let mut counter = 0;

        for element in &builder.uniform_buffer_info {
            Self::ubo_to_bind(element, &mut bindings[counter]);
            counter += 1;
        }
        for element in &builder.samplers_info {
            Self::sampler_to_bind(element, &mut bindings[counter]);
            counter += 1;
        }
        for item in &builder.acceleration_structures_info {
            Self::as_to_bind(item, &mut bindings[counter]);
            counter += 1;
        }
        for item in &builder.storage_images_info {
            Self::storage_image_to_bind(item, &mut bindings[counter]);
            counter += 1;
        }
        for item in &builder.storage_buffer_infos {
            Self::sbo_to_bind(item, &mut bindings[counter]);
            counter += 1;
        }

        if !bindings.is_empty() {
            let layout_info = vk::DescriptorSetLayoutCreateInfo::default()
                .bindings(&bindings);

            self.descriptor_set_layout = Some(
                unsafe {
                    self.device
                        .create_descriptor_set_layout(&layout_info, None)
                        .expect("Failed to create descriptor set layout")
                }
            );
        }
    }

    fn load_pipeline_layout(&mut self, builder: &VlPipelineBuilder) {
        let mut push_constant_ranges = vec![vk::PushConstantRange::default(); builder.push_constant_infos.len()];
        for (i, element) in builder.push_constant_infos.iter().enumerate() {
            Self::info_to_range(element, &mut push_constant_ranges[i]);
        }

        let layout_info = vk::PipelineLayoutCreateInfo::default()
            .set_layouts(self.descriptor_set_layout.as_ref().map_or(&[][..], |dsl| std::slice::from_ref(dsl)))
            .push_constant_ranges(&push_constant_ranges);

        self.pipeline_layout = unsafe {
            self.device
                .create_pipeline_layout(&layout_info, None)
                .expect("Failed to create pipeline layout")
        };
    }

    fn prepare_binding(&mut self, inputs: &[VlVertexInput]) {
        let stride: usize = inputs.iter().map(|i| i.type_size * i.coordinates_amount as usize).sum();
        self.input_bind_desc.binding = 0;
        self.input_bind_desc.stride = stride as u32;
        self.input_bind_desc.input_rate = vk::VertexInputRate::VERTEX;
    }

    fn prepare_input_attribs(&mut self, inputs: &[VlVertexInput]) {
        self.input_attrib_descs.resize(inputs.len(), vk::VertexInputAttributeDescription::default());
        let mut offset_count = 0;

        for (i, element) in inputs.iter().enumerate() {
            self.input_attrib_descs[i].binding = 0;
            self.input_attrib_descs[i].location = element.location;
            self.input_attrib_descs[i].offset = offset_count as u32;
            self.input_attrib_descs[i].format = element.format;
            offset_count += element.type_size * element.coordinates_amount as usize;
        }
    }

    fn info_to_range(info: &VlPushConstantInfo, range: &mut vk::PushConstantRange) {
        range.stage_flags = info.shader_stages;
        range.offset = 0;
        range.size = info.size as u32;
    }

    fn ubo_to_bind(info: &VlUniformBufferInfo, binding: &mut vk::DescriptorSetLayoutBinding) {
        binding.binding = info.binding;
        binding.descriptor_type = vk::DescriptorType::UNIFORM_BUFFER;
        binding.descriptor_count = info.descriptor_count;
        binding.stage_flags = info.shader_stages;
    }

    fn sbo_to_bind(info: &VlStorageBufferInfo, binding: &mut vk::DescriptorSetLayoutBinding) {
        binding.binding = info.binding;
        binding.descriptor_type = vk::DescriptorType::STORAGE_BUFFER;
        binding.descriptor_count = info.descriptor_count;
        binding.stage_flags = info.stage_flags;
    }

    fn sampler_to_bind(info: &VlSamplerInfo, binding: &mut vk::DescriptorSetLayoutBinding) {
        binding.binding = info.binding;
        binding.descriptor_type = vk::DescriptorType::COMBINED_IMAGE_SAMPLER;
        binding.descriptor_count = info.descriptor_count;
        binding.stage_flags = info.shader_stages;
    }

    fn as_to_bind(info: &VlAccelerationStructureInfo, binding: &mut vk::DescriptorSetLayoutBinding) {
        binding.binding = info.binding;
        binding.descriptor_type = vk::DescriptorType::ACCELERATION_STRUCTURE_KHR;
        binding.descriptor_count = info.descriptor_count;
        binding.stage_flags = info.shader_stages;
    }

    fn storage_image_to_bind(info: &VlStorageImageInfo, binding: &mut vk::DescriptorSetLayoutBinding) {
        binding.binding = info.binding;
        binding.descriptor_type = vk::DescriptorType::STORAGE_IMAGE;
        binding.descriptor_count = info.descriptor_count;
        binding.stage_flags = info.shader_stages;
    }

    pub unsafe fn destroy(&mut self) {
        if self.pipeline_layout != vk::PipelineLayout::null() {
            self.device.destroy_pipeline_layout(self.pipeline_layout, None);
        }
        if let Some(layout) = self.descriptor_set_layout.take() {
            self.device.destroy_descriptor_set_layout(layout, None);
        }
    }

    pub fn pipeline_layout(&self) -> vk::PipelineLayout {
        self.pipeline_layout
    }

    pub fn descriptor_set_layout(&self) -> Option<vk::DescriptorSetLayout> {
        self.descriptor_set_layout
    }

    pub fn input_bind_desc(&self) -> vk::VertexInputBindingDescription {
        self.input_bind_desc
    }

    pub fn input_attrib_descs(&self) -> &Vec<vk::VertexInputAttributeDescription> {
        &self.input_attrib_descs
    }

}
